backends/k8s/client.go

```diff
@@ -106,212 +117,176 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
+		return err
+	}
+	kvps = append(kvps, cNode)
+
+	// Add per-node BGP config (each of the per-node resource clients also implements
+	// the CustomK8sNodeResourceList interface, used to extract per-node resources from
+	// the Node resource.
+	if cfg, err := c.nodeBgpCfgClient.ExtractResourcesFromNode(kNode); err != nil {
+		log.Error("Failed to parse BGP configs from node resource - skip config data")
+	} else {
+		kvps = append(kvps, cfg...)
 	}
 
-	switch prefix {
-	case global:
-		// These are currently not changeable in k8s.
-		time.Sleep(10 * time.Second)
-		return waitIndex, nil
-	case globalNodeMesh:
-		// This is currently not changeable in k8s.
-		time.Sleep(10 * time.Second)
-		return waitIndex, nil
-	case allNodes:
-		w, err := c.clientSet.Nodes().Watch(kapiv1.ListOptions{})
-		if err != nil {
-			return waitIndex, err
-		}
-		event := <-w.ResultChan()
-		ver := event.Object.(*kapiv1.NodeList).ListMeta.ResourceVersion
-		w.Stop()
-		log.Debug(fmt.Sprintf("%d : %s", waitIndex, ver))
-
-		return convertResourceVersionToUint(ver, prefix)
-	case ipPool:
-		w, err := c.tprClient.Get().
-			Name("ippool").
-			Namespace("kube-system").
-			Watch()
-		if err != nil {
-			return waitIndex, err
-		}
-		event := <-w.ResultChan()
-		ver := event.Object.(*thirdparty.IpPoolList).Metadata.ResourceVersion
-		w.Stop()
-
-		return convertResourceVersionToUint(ver, prefix)
-	default:
-		// We aren't tracking this key, default to 10 second refresh.
-		time.Sleep(60 * time.Second)
-		log.Debug(fmt.Sprintf("Receieved unknown key: %v", prefix))
-		return waitIndex + 1, nil
+	if peers, err := c.nodeBgpPeerClient.ExtractResourcesFromNode(kNode); err != nil {
+		log.Error("Failed to parse BGP peers from node resource - skip config data")
+	} else {
+		kvps = append(kvps, peers...)
 	}
-	return waitIndex, nil
+
+	// Populate the vars map from the KVPairs.
+	c.populateFromKVPairs(kvps, vars)
+
+	return nil
 }
 
-// buildTPRClient builds a RESTClient configured to interact with Calico ThirdPartyResources.
-func buildTPRClient(baseConfig *rest.Config) (*rest.RESTClient, error) {
-	// Generate config using the base config.
-	cfg := baseConfig
-	cfg.GroupVersion = &schema.GroupVersion{
-		Group:   "projectcalico.org",
-		Version: "v1",
+// populateFromKVPairs populates the vars KV map from the supplied set of
+// KVPairs.  This uses the libcalico-go compat module and serialization functions
+// to write out the KVPairs in etcdv2 format.  This works in conjunction with the
+// etcdVarClient defined below which provides a "mock" etcd backend which actually
+// just writes out data to the vars map.
+func (c *Client) populateFromKVPairs(kvps []*model.KVPair, vars map[string]string) {
+	// Create a etcdVarClient to write the KVP results in the vars map, using the
+	// compat adaptor to write the values in etcdv2 format.
+	client := compat.NewAdaptor(&etcdVarClient{vars: vars})
+	for _, kvp := range kvps {
+		client.Apply(kvp)
```

Comment:
Should we be handling the errors returned by this?  I think we need to at least log.